source/common/buffer/watermark_buffer.cc

```diff
@@ -72,12 +72,16 @@ void WatermarkBuffer::setWatermarks(uint32_t low_watermark, uint32_t high_waterm
   ASSERT(low_watermark < high_watermark || (high_watermark == 0 && low_watermark == 0));
   low_watermark_ = low_watermark;
   high_watermark_ = high_watermark;
-  checkHighWatermark();
+  overflow_watermark_ = high_watermark * overflow_multiplier_;
+  // TODO(adip): What should be done if there's an overflow (overflow_watermark_ <
+  // overflow_watermark_)? should this be a release_assert?
+  ASSERT((high_watermark <= overflow_watermark_) || (overflow_multiplier_ == 0));
+  checkHighAndOverflowWatermarks();
   checkLowWatermark();
 }
 
 void WatermarkBuffer::checkLowWatermark() {
-  if (!above_high_watermark_called_ ||
+  if (above_overflow_watermark_called_ || !above_high_watermark_called_ ||
```

Comment:
Why do we check for above_overflow_watermark_called_ ?

Reply:
At the moment, although an overflow occurs, the buffer can still be used. Once the buffer is read from, the low watermark callback will be called. I'm not sure that in cases where the connection is terminated, the `readDisable` should be called again.
If I'm missing a scenario please let me know.

That said, a different buffer design that should be considered is to disable all operations once an overflow has been detected.

Reply:
Here's the problem: if the overflow callback is called, the buffer stops functioning properly since low watermark callbacks are no longer called when they should.

This PR feels high risk.  I think we may be jumping the gun by going straight to closing the connection/resetting stream on overflow.  It would be good to gather some data about what can cause the overflow cb to trigger.  I know there's a runtime flag to control this behavior, but it defaulting on with a 2x overflow ratio could cause some problems.

Reply:
Agreed that runtime control over the high watermark constant seems a useful capability.

Reply:
@yanavlasov triggering the 2x multipler overflow watermark is very easy.  Specially when you factor in that the H2 connection flow control window is not coordinated with high watermarks.  Going straight to closing connection is very risky.

I'm just arguing for keeping the high/low watermarks in working order even if the overflow one is hit, so we can add overflow implementations that just increment counters.